### PR TITLE
[IM] Remove CommandSender object pool in interaction model engine

### LINKED ISSUE #8958: C1: Fix-up CHIPClusters.cpp to internally use `SingleInvokeInitiator`
#### Problem
Fix the generated logic in `CHIPClusters.cpp` to internally use `SingleInvokeInitiator` instead of `CommandSender` _without_ changing its public APIs.

This permits validation of the new `SingleInvokeInitiator` class without actually changing the various app call-sites.

As part of this fix, the internal guts of that logic should now not have to use any of the gCallbackMgr logic.

### PR DESCRIPTION
**Pull request by erjiaqing**
#### Problem

Fixes #8958
CommandSender part of #9460

Use app provided allocator for CommandSender objects.

#### Change overview

- Introduce CommandSenderDelegate, current as a deleter, will move callbacks like `CommandResponseProcessed` to this delegate when implementing https://github.com/project-chip/connectedhomeip/issues/8960.
- Removed Shutdown(), the destructor is Shutdown() for CommandSender.
- Update test cases
- Add allocator and destructor in the IM delegate of CHIP Device Controller.
- Moved CommandSender related callbacks to CommandSender::Delegate, CommandHandler and Command base class no longer requires a delegate.
- Run codegen
- No CHIPCluster API changes.

#### Testing
- Already covered, test passed.

### CODE REVIEW

File: src/app/CommandSender.h [hunk 1]
```diff
@@ -36,15 +36,69 @@
...
 public:
+    // TODO: Move various CommandResponse related callbacks to this delegate.
+    class Delegate
+    {
+    public:
+        virtual ~Delegate() = default;
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received, aStatus will be filled with the content of the StatusElement, aData will be a
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:58]: I'm not certain this is the API pattern we want.

In most parts of the code-base, there are usually pairs of callbacks defined for any given fail-able operation: a success callback and a failure callback.

Receipt of non-empty data, empty-data, or a status element with success in it should all be dispatched through this `OnResponse` callback. status elements with failure codes I don't think should be routed through this callback, since it really makes it inconsistent with the patterns in the rest of the code-base.

File: src/app/CommandSender.h [hunk 2]
```diff
@@ -36,15 +36,69 @@
...
 
 namespace chip {
 namespace app {
 
-class CommandSender : public Command, public Messaging::ExchangeDelegate
+class CommandSender final : public Command, public Messaging::ExchangeDelegate
 {
 public:
+    // TODO: Move various CommandResponse related callbacks to this delegate.
+    class Delegate
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:51]: The term 'delegate' is way too over-used in the SDK, and doesn't convey precisely what this interface is.

Suggestion:
```suggestion
    class ResponseHandlerInterface
```

File: src/app/CommandSender.h [hunk 3]
```diff
@@ -36,15 +36,69 @@
...
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction, MUST NOT be freed in this
+         * callback.
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aStatus: The status element, filled with actual StatusElement when a StatusElement is received, when command
+         * specific response is received, aStatus will always be filled with success status code.
+         * @param[in] aData: The command data, or a uninitialized TLVReader if a StatusElement is received.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath,
+                                const StatusElement::Type & aStatus, TLV::TLVReader & aData)
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:71]: Given the above comment, `aStatus` should not need to be passed in here.

Instead, convert the TLVReader reference to a pointer, with a null pointer indicating the absence of any received data, and instead, just a success status element received instead.

File: src/app/CommandSender.h [hunk 4]
```diff
@@ -36,15 +36,69 @@
...
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
+         * - TLV Related erros: The CommandSender received a malformed response from server.
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction, MUST NOT be freed in this
+         * callback.
+         * @param[in] aError: The error occurred during the invoke command interaction.
+         */
+        virtual void OnError(const CommandSender * apCommandSender, CHIP_ERROR aError) {}
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:83]: We should be conveying the IM status codes as well as cluster-specific status codes (if one was present) through this API.

Please file a follow-up issue to fix this.

File: src/app/CommandSender.h [hunk 5]
```diff
@@ -36,15 +36,69 @@
...
+
+        /**
+         * ReleaseCommandSender will be called when CommandSender has finished all the works, and it is safe to release the
+         * allocated CommandSender object.
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:87]:
```suggestion
         * allocated CommandSender object if needed. 
```

File: src/app/CommandSender.h [hunk 6]
```diff
@@ -36,15 +36,69 @@
...
+         *
+         * This function is marked as must be implemented to make application aware of releasing the object.
+         *
+         * This function will always be called eaxctly once for a single command sender.
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:91]:
```suggestion
         * This function will always be called exactly once for a single command sender.
         *
         * This object cannot be used again till Init has been called again on this object.
```

File: src/app/Command.h [hunk 7]
```diff
@@ -75,13 +75,7 @@ class Command
...
      *
      */
-    CHIP_ERROR Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate);
-
-    /**
-     *  Shutdown the Command. This terminates this instance
-     *  of the object and releases all held resources.
-     */
-    void Shutdown();
+    CHIP_ERROR Init(Messaging::ExchangeManager * apExchangeMgr);
```
Comment (mrjerryjohns) [inline src/app/Command.h:78]: This removes the `InteractionModelDelegate` on the server side too - does this cause any unintended consequences?

File: src/app/MessageDef/CommandPath.h [hunk 8]
```diff
@@ -44,6 +44,16 @@ enum
     kCsTag_CommandId  = 3,
 };
 
+struct Type
+{
+    chip::EndpointId endpointId;
+    chip::ClusterId clusterId;
+    chip::CommandId commandId;
```
Comment (mrjerryjohns) [inline src/app/MessageDef/CommandPath.h:51]: What about the rest of the metadata in a command, like SuppressResponse?

What about paths targeting groups?

Since that isn't the focus of this PR, please file a follow-up issue for the two items.

File: src/controller/CHIPDevice.cpp [hunk 9]
```diff
@@ -158,6 +160,31 @@ CHIP_ERROR Device::SendCommands(app::CommandSender * commandObj)
     return commandObj->SendCommandRequest(mDeviceId, mFabricIndex, mSecureSession);
 }
 
+CHIP_ERROR Device::NewCommandSender(app::CommandSender ** commandSender)
```
Comment (mrjerryjohns) [inline src/controller/CHIPDevice.cpp:163]: I don't see the value of having this little wrapper to allocate a `CommandSender`, which in turn, calls into a delegate to allocate one.

Callers should just allocate a CommandSender in the way they see fit (std::make_unique, new, on stack).

File: zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp [hunk 10]
```diff
@@ -138,7 +138,7 @@ CHIP_ERROR AccountLoginCluster::Login(Callback::Cancelable * onSuccessCallback,
     app::CommandPathParams cmdParams = { mEndpoint, /* group id */ 0, mClusterId, AccountLogin::Commands::Ids::Login,
                                          (app::CommandPathFlags::kEndpointIdValid) };
 
-    SuccessOrExit(err = app::InteractionModelEngine::GetInstance()->NewCommandSender(&sender));
+    SuccessOrExit(err = mDevice->NewCommandSender(&sender));
```
Comment (mrjerryjohns) [inline zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp:141]: Like I mentioned earlier, just call `new` or Platform::New directly here. Let's reduce the # of layers here and simplify.

File: src/app/Command.h (hunk 7 above)
Comment (erjiaqing) [reply src/app/Command.h:78]: After going through the code, CommandHandler does not use IMDelegate.

Regarding #9460, future callbacks will using seperate callback class.

Comment (erjiaqing) [reply src/app/Command.h:78]: After going through the code, CommandHandler does not use IMDelegate.

Regarding #9460, future callbacks will using seperate callback class.

File: src/app/CommandSender.h (hunk 1 above)
Comment (erjiaqing) [reply src/app/CommandSender.h:58]: I'm also thinking about this, in my mind, the result can be 
Success, Expected Error and Unexpected Error
The Expected Error can be interaction model protocol code (i.e. Command not found, then server replied with the state code)
And Command specific error codes, we cannot deliver this to OnError since the actual type varies from different commands / clusters.

The unexpected error is the server size misbehaved, e.g. network error, / malformed response (bad TLV) etc.

I'm not certain whether a client will expect a failure status code in im when sending commands, if this is not true, then we should put im error with protocol error handling into seperate callback.

I'm also thinking about some other behaviors, i.e. HTTP requests, seems most library will regard HTTP error as exceptions, so maybe we should call OnError on interaction model errors.

**Comment by erjiaqing** [comment]
@bzbarsky-apple @vivien-apple PTAL

File: src/app/CommandSender.h [hunk 11]
```diff
@@ -36,15 +38,71 @@
...
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aData: The command data, will be nullptr if the server returns a StatusElement.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath, TLV::TLVReader * aData) {}
+
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
+         * - TLV Related erros: The CommandSender received a malformed response from server.
+         * - CHIP_NO_ERROR: The CommandSender received a legal response from server, but the protocol code indicates a error status
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:75]: It's a little counter-intuitive to receive a CHIP_NO_ERROR (which indicates success), but still receive a protocol status code indicating an error.

I'd rather we return a CHIP error like (CHIP_IM_ERROR) that more clearly indicates that there is a more detailed error code that will convey the status.

In the future, we really should have IM be a sub-component in the `SdkPart` within the CHIP error space, allowing us to fold this down to just CHIP_ERROR.

File: src/app/CommandSender.h [hunk 12]
```diff
@@ -36,15 +38,71 @@
...
+         * OnFinal will be called when CommandSender has finished all the works, and it is safe to release the
+         * allocated CommandSender object if needed.
+         *
+         * This function is marked as must be implemented to make application aware of releasing the object.
+         *
+         * This function will always be called exactly once for a single command sender.
+         *
+         * @param[in] apCommandSender: The command sender object of the terminated invoke command transaction.
+         */
+        virtual void OnFinal(CommandSender * apCommandSender) = 0;
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:99]: Needs clarification on whether this will be called upon encountering errors calling into any of the public APIs on `CommandSender` object, or whether this will only be expected to be called after calling `SendCommandRequest`.

Comment (erjiaqing) [reply src/app/CommandSender.h:99]: Done

File: src/app/tests/TestCommandInteraction.cpp [hunk 13]
```diff
@@ -87,11 +87,29 @@ void DispatchSingleClusterCommand(chip::ClusterId aClusterId, chip::CommandId aC
 void DispatchSingleClusterResponseCommand(chip::ClusterId aClusterId, chip::CommandId aCommandId, chip::EndpointId aEndPointId,
                                           chip::TLV::TLVReader & aReader, CommandSender * apCommandObj)
 {
-    ChipLogDetail(Controller, "Received Cluster Command: Cluster=%" PRIx32 " Command=%" PRIx32 " Endpoint=%" PRIx16, aClusterId,
-                  aCommandId, aEndPointId);
     // Nothing todo.
 }
 
+class MockCommandSenderCallback : public CommandSender::Callback
```
[the remark this answers is not in the document]
Comment (erjiaqing) [reply src/app/tests/TestCommandInteraction.cpp:114]: Done
2 is not expected since we always expect a status code or payload.

The leaked exchange is checked after running all the tests.

File: src/app/Command.cpp [hunk 14]
```diff
@@ -33,15 +33,14 @@
 namespace chip {
 namespace app {
 
-CHIP_ERROR Command::Init(Messaging::ExchangeManager * apExchangeMgr, InteractionModelDelegate * apDelegate)
+CHIP_ERROR Command::Init(Messaging::ExchangeManager * apExchangeMgr)
```
Comment (msandstedt) [inline src/app/Command.cpp:36]: It seems very non-symmetrical to remove Shutdown, but retain Init.

Why can't the body of Init be moved to the Command constructor?

Comment (msandstedt) [reply src/app/Command.cpp:36]: Oh, I see.  Shutdown is moved to the inheriting classes.

Would it be cleaner though to still retain Shutdown in the base Command class for it to release the resources it owns?  Inheriting classes can release things specific to their extended implementation and the call the base class Shutdown.

It's just confusing to see the setup and teardown of Command spread between itself and the classes that inherit it.

File: src/app/MessageDef/StatusElement.h [hunk 15]
```diff
@@ -45,6 +45,13 @@ enum
     kCsTag_ClusterId    = 4
 };
 
+struct Type
```
Comment (msandstedt) [inline src/app/MessageDef/StatusElement.h:48]: I don't know what the 'Element' suffix means, but I see it is ubiquitous in 'src/app/MessageDef'.  But by the time we get to this struct, we have: `chip::app::StatusElement::Type`.  That's a bit confusing when what this actually is is a StatusReport, isn't it?  Can we just call it that?

File: src/app/CommandSender.cpp [hunk 16]
```diff
@@ -77,20 +77,15 @@ CHIP_ERROR CommandSender::OnMessageReceived(Messaging::ExchangeContext * apExcha
...
+    if (mpCallback != nullptr)
     {
         if (err != CHIP_NO_ERROR)
         {
-            mpDelegate->CommandResponseError(this, err);
-        }
-        else
-        {
-            mpDelegate->CommandResponseProcessed(this);
+            mpCallback->OnError(this, Protocols::InteractionModel::Status::Failure, err);
```
Comment (msandstedt) [inline src/app/CommandSender.cpp:88]: Do these changes mean we only get a callback on error and not success?  That seems like a problem.

File: src/app/MessageDef/CommandPath.h (hunk 8 above)
Comment (msandstedt) [inline src/app/MessageDef/CommandPath.h:51]: It would be nice if these member variables had the m prefix (which I think is our general convention) so that it would be obvious that they are member variables in src/app/MessageDef/CommandPath.cpp.

```suggestion
    chip::EndpointId mEndpointId;
    chip::ClusterId mClusterId;
    chip::CommandId mCommandId;
```

File: src/app/MessageDef/CommandPath.h [hunk 17]
```diff
@@ -44,6 +44,16 @@ enum
     kCsTag_CommandId  = 3,
 };
 
+struct Type
```
Comment (msandstedt) [inline src/app/MessageDef/CommandPath.h:47]: The word Type is just too overloaded to be intuitive here.   I suppose we are saying that this is the 'type' of the path.  But I think it would be a lot clearer to just call this 'Path' or similar.

File: src/controller/CHIPDeviceController.cpp [hunk 18]
```diff
@@ -1792,55 +1792,40 @@ void DeviceCommissioner::OnNodeDiscoveryComplete(const chip::Mdns::DiscoveredNod
...
-    return CHIP_NO_ERROR;
+    if (aData != nullptr)
+    {
+        // TODO: We are using const_cast, this line will be removed by implementing #8971
+        chip::app::DispatchSingleClusterResponseCommand(aPath.clusterId, aPath.commandId, aPath.endpointId, *aData,
+                                                        const_cast<chip::app::CommandSender *>(apCommandSender));
+    }
+    else
+    {
+        IMDefaultResponseCallback(apCommandSender, EMBER_ZCL_STATUS_SUCCESS);
```
Comment (msandstedt) [inline src/controller/CHIPDeviceController.cpp:1811]: It appears this theoretically could give a kind of backward compatibility if adata is null.  However, this is pretty incomplete.  We're burying any potential errors and always reporting success.

Under what conditions can adata be null?  Do we need to retain this legacy behavior?

Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:1811]: Ah, I think I see.  By virtue of landing in OnResponse, we have by definition been successful.

But my other question still stands.  What conditions will cause adata to be null?  Is the IMDefaultResponseCallback a legacy artifact?  It appears it is.  Do we need to keep it?  What does it give us?

Comment (msandstedt) [reply src/controller/CHIPDeviceController.cpp:1811]: OK, seems to be answered in `src/app/CommandSender.h`:

```
OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received and it is Success, aData will be nullptr.
+         *  - When command specific response is received, aData will be a valid TLVReader initialized with a struct of command
```

Makes sense.

File: src/app/CommandSender.h [hunk 19]
```diff
@@ -36,15 +38,71 @@
...
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction continue to exist after
+         * this call completes.
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aData: The command data, will be nullptr if the server returns a StatusElement.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath, TLV::TLVReader * aData) {}
+
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
```
Comment (msandstedt) [inline src/app/CommandSender.h:72]:
```suggestion
         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive an expected response.
```

File: src/app/CommandSender.h [hunk 20]
```diff
@@ -36,15 +38,71 @@
...
+         * After the command being sent, this function will always be called exactly once for a single command sender.
+         *
+         * @param[in] apCommandSender: The command sender object of the terminated invoke command transaction.
+         */
+        virtual void OnFinal(CommandSender * apCommandSender) = 0;
+    };
+
+    CommandSender(Callback * apCallback) : mpCallback(apCallback){};
+
+    ~CommandSender() { AbortExistingExchangeContext(); }
```
Comment (mrjerryjohns) [inline src/app/CommandSender.h:104]: The actual exchange context is 'owned' by the base class, so it feels more appropriate to abort it in `~Command()`.

File: src/controller/DeviceControllerInteractionModelDelegate.h [hunk 21]
```diff
@@ -10,20 +10,20 @@ namespace Controller {
...
+    void OnError(const app::CommandSender * apCommandSender, Protocols::InteractionModel::Status aInteractionModelStatus,
+                 CHIP_ERROR aProtocolError) override;
+    void OnFinal(app::CommandSender * apCommandSender) override;
 
-    CHIP_ERROR CommandResponseProtocolError(const app::CommandSender * apCommandSender, uint8_t aCommandIndex) override;
-
-    CHIP_ERROR CommandResponseError(const app::CommandSender * apCommandSender, CHIP_ERROR aError) override;
-
-    CHIP_ERROR CommandResponseProcessed(const app::CommandSender * apCommandSender) override;
+    app::CommandSender * NewCommandSender();
```
Comment (mrjerryjohns) [inline src/controller/DeviceControllerInteractionModelDelegate.h:26]: Why is this here? Also, no one seems to be calling this.

File: src/lib/core/CHIPError.h [hunk 22]
```diff
@@ -2198,6 +2198,15 @@ using CHIP_ERROR = ::chip::ChipError;
...
 #define CHIP_ERROR_NO_SHARED_TRUSTED_ROOT                      CHIP_CORE_ERROR(0xc9)
 
+/*
+ * @def CHIP_ERROR_IM
+ *
+ * @brief
+ *   General error code when a interaction model error is received, usually
+ * comes with a interaction model status code beside.
+ */
+#define CHIP_ERROR_IM                                          CHIP_CORE_ERROR(0xca)
```
Comment (mrjerryjohns) [inline src/lib/core/CHIPError.h:2208]: CHIP_ERROR_IM_ERROR_RECEIVED might be a better name.

File: zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp [hunk 23]
```diff
@@ -162,8 +164,11 @@ CHIP_ERROR AdministratorCommissioningCluster::OpenBasicCommissioningWindow(Callb
                                          AdministratorCommissioning::Commands::Ids::OpenBasicCommissioningWindow,
                                          (app::CommandPathFlags::kEndpointIdValid) };
 
-    SuccessOrExit(err = app::InteractionModelEngine::GetInstance()->NewCommandSender(&sender));
+    CommandSenderHandler sender(Platform::New<app::CommandSender>(mDevice->GetInteractionModelDelegate()));
```
Comment (mrjerryjohns) [inline zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp:167]: I'm still not certain why we need to use `Platform::New` and a custom deleter. Why not just `make_unique`?

File: src/controller/CHIPDeviceController.cpp [hunk 24]
```diff
@@ -1792,55 +1792,40 @@ void DeviceCommissioner::OnNodeDiscoveryComplete(const chip::Mdns::DiscoveredNod
...
-    // Note: The IMDefaultResponseCallback is a bridge to the old CallbackMgr before IM is landed, so it still accepts EmberAfStatus
-    // instead of IM status code.
-    IMDefaultResponseCallback(apCommandSender, EMBER_ZCL_STATUS_FAILURE);
-
-    return CHIP_NO_ERROR;
+    if (aData != nullptr)
+    {
+        // TODO: We are using const_cast, this line will be removed by implementing #8971
+        chip::app::DispatchSingleClusterResponseCommand(aPath.clusterId, aPath.commandId, aPath.endpointId, *aData,
+                                                        const_cast<chip::app::CommandSender *>(apCommandSender));
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1807]: How about we not have the const_cast here and have our API match reality (i.e. _not_ take a const CommandSender, and then if we ever get to a point where we not doing non-const things with it anymore _then_ we change the API to take const.

File: src/controller/CHIPDeviceController.cpp [hunk 25]
```diff
@@ -1792,55 +1792,40 @@ void DeviceCommissioner::OnNodeDiscoveryComplete(const chip::Mdns::DiscoveredNod
...
 {
     // Generally IM has more detailed errors than ember library, here we always use EMBER_ZCL_STATUS_FAILURE before #6308 is landed
     // and the app can take care of these error codes, the actual handling of the commands should implement full IMDelegate.
     // #6308: By implement app side IM delegate, we should be able to accept detailed error codes.
     // Note: The IMDefaultResponseCallback is a bridge to the old CallbackMgr before IM is landed, so it still accepts EmberAfStatus
     // instead of IM status code.
-    IMDefaultResponseCallback(apCommandSender, EMBER_ZCL_STATUS_FAILURE);
-
-    return CHIP_NO_ERROR;
+    IMDefaultResponseCallback(apCommandSender, app::ToEmberAfStatus(aProtocolCode));
```
Comment (bzbarsky-apple) [inline src/controller/CHIPDeviceController.cpp:1823]: First, this code no longer matches the comment above it; fix the comment.

Second, at least document why we are ignoring aError.

File: src/controller/CommandSenderAllocator.h [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+namespace chip {
+namespace Controller {
+
+class CommandSenderPlatformDeleter
+{
+public:
+    void operator()(app::CommandSender * commandSender) const { chip::Platform::Delete(commandSender); }
+};
+
+using CommandSenderHandler = std::unique_ptr<app::CommandSender, CommandSenderPlatformDeleter>;
```
Comment (bzbarsky-apple) [inline src/controller/CommandSenderAllocator.h:35]:
```suggestion
using CommandSenderHandle = std::unique_ptr<app::CommandSender, CommandSenderPlatformDeleter>;
```

File: src/controller/python/chip/interaction_model/Delegate.cpp [hunk 27]
```diff
@@ -32,52 +32,43 @@ namespace Controller {
...
+void PythonInteractionModelDelegate::OnResponse(const app::CommandSender * apCommandSender, const app::CommandPath::Type & aPath,
+                                                TLV::TLVReader * aData)
 {
-    CommandStatus status{ aProtocolId, aProtocolCode, aEndpointId, aClusterId, aCommandId, aCommandIndex };
+    CommandStatus status{ Protocols::InteractionModel::Id.ToFullyQualifiedSpecForm(),
+                          to_underlying(Protocols::InteractionModel::Status::Success),
+                          aPath.endpointId,
+                          aPath.clusterId,
+                          aPath.commandId,
+                          1 };
```
Comment (bzbarsky-apple) [inline src/controller/python/chip/interaction_model/Delegate.cpp:43]: What is this "1" here?  Please at least document, if not come up with a better name for it.

File: src/app/CommandSender.h [hunk 28]
```diff
@@ -36,15 +38,71 @@
...
+         * @param[in] aPath: The command path field in invoke command response.
+         * @param[in] aData: The command data, will be nullptr if the server returns a StatusElement.
+         */
+        virtual void OnResponse(const CommandSender * apCommandSender, const CommandPath::Type & aPath, TLV::TLVReader * aData) {}
+
+        /**
+         * OnError will be called when error occurred when receiving / handling response from server:
+         * - CHIP_ERROR_TIMEOUT: The CommandSender failed to receive a expected response.
+         * - TLV Related erros: The CommandSender received a malformed response from server.
+         * - CHIP_ERROR_IM: The CommandSender received a legal response from server, but the protocol code indicates a error status
```
[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply src/app/CommandSender.h:74]: 1. Servers today send back just two error types: 'secure channel general status code' and an IM status code. In all cases, the IM status code conveys a status code back from the server for errors it encounters. This is conveyed through the aInteractionModleStatus arg and is only to be interpreted when when aProtocolError == CHIP_ERROR_IM (as described in this line).
2. See line above, i.e "CommandSender receives a malformed response from server".
3. See 2 lines above, "CommandSender failed to receive an expected response".
4. aProtocolError != CHIP_NO_ERROR, as is the case in general.

File: src/app/MessageDef/CommandPath.cpp [hunk 29]
```diff
@@ -30,12 +30,48 @@
...
+namespace CommandPath {
+CHIP_ERROR Type::Decode(chip::TLV::TLVReader & reader)
+{
+    CHIP_ERROR err;
+    TLV::TLVType outer;
+
+    err = reader.EnterContainer(outer);
+    ReturnErrorOnFailure(err);
+
+    while ((err = reader.Next()) == CHIP_NO_ERROR)
```
[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply src/app/MessageDef/CommandPath.cpp:51]: Even if you were to write the code in a way that expects that order, it'll still end up looking like this, because of the fact that these tags _may_ not be present.

At this point, adding code to enforce ordering will be extra and only serves to make this less flexible.

File: src/app/MessageDef/CommandPath.h (hunk 17 above)
[the remark this answers is not in the document]
Comment (mrjerryjohns) [reply src/app/MessageDef/CommandPath.h:47]: I think there is an incorrect modeling of this as a 'data model type' (i.e cluster object type). Agreed that it should just be a vanilla `CommandPath`.

File: zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp (hunk 23 above)
Comment (bzbarsky-apple) [reply zzz_generated/controller-clusters/zap-generated/CHIPClusters.cpp:167]: To make sure we use the right heap?

File: src/app/Command.cpp (hunk 14 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/Command.cpp:36]: Well, we have split init no matter what, in that the IM delegate is passed to the constructor.

I do question the value the Init method, by the way: the allocations it does could happen when someone tries to actually do something that needs the memory, instead of requiring up-front init...

> I think the pattern in this PR of not permitting re-use (i.e Init can only be called once ever for a given object)

I don't think this PR does that.  It does require a Shutdown and new Init before you can reuse the object.

> If I'm an embedded device that wants to keep re-using the same CommandSender object (which extends this) to send multiple commands in sequence

You can't re-use it until the response to the first command returns, right?  And then you can just destroy and re-create the object, I'd think.  That is, you can allocate memory for the object and then placement-new + delete as needed into that memory, if you want to keep reusing the memory.

So as a followup to this PR I think we should pretty strongly consider requiring that you create this object once, use it, then destroy it.  It makes the lifetime bits very clear.

File: src/app/CommandSender.cpp [hunk 30]
```diff
@@ -114,55 +109,68 @@ void CommandSender::OnResponseTimeout(Messaging::ExchangeContext * apExchangeCon
...
-    err = aCommandElement.GetStatusElement(&statusElementParser);
-    if (CHIP_NO_ERROR == err)
     {
-        err = statusElementParser.DecodeStatusElement(&generalCode, &protocolId, &protocolCode);
-        SuccessOrExit(err);
-        if (mpDelegate != nullptr)
+        bool hasCommandSpecificResponse = false;
+        chip::TLV::TLVReader commandDataReader;
+
+        // Default to success when command specify response is received.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.cpp:126]: Thank you, that is clearer.  The `hasCommandSpecificResponse` variable should perhaps also be renamed to `hasDataResponse` or something?

File: src/app/CommandSender.cpp [hunk 31]
```diff
@@ -114,55 +109,68 @@ void CommandSender::OnResponseTimeout(Messaging::ExchangeContext * apExchangeCon
...
-    {
-        // TODO(Spec#3258): The endpoint id in response command is not clear, so we cannot do "ClientClusterCommandExists" check.
-        err = aCommandElement.GetData(&commandDataReader);
         SuccessOrExit(err);
-        // TODO(#4503): Should call callbacks of cluster that sends the command.
-        DispatchSingleClusterResponseCommand(clusterId, commandId, endpointId, commandDataReader, this);
+
+        if (mpCallback != nullptr)
+        {
+            if (statusElement.protocolCode == to_underlying(Protocols::InteractionModel::Status::Success))
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.cpp:145]: > See CommandHandler::AddStatusCode:

Nothing guarantees that the other side used that, or indeed used this SDK.  They could have just sent garbage data.

I agree that fixing status code stuff here in general should be a separate PR, but I don't think we should be introducing obviously-buggy code that can be avoided by actually checking the protocol id in `statusElement`.

File: src/app/CommandSender.h [hunk 32]
```diff
@@ -36,15 +38,71 @@
...
+    class Callback
+    {
+    public:
+        virtual ~Callback() = default;
+
+        /**
+         * OnResponse will be called when the response from server arrived.
+         *  - When StatusElement is received and it is Success, aData will be nullptr.
+         *  - When command specific response is received, aData will be a valid TLVReader initialized with a struct of command
+         * specify response.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.h:61]: I am not asking us to fix the implementation.  If we can't write sane API docs until we fix the impl, so be it; if we can make the docs sensical immediately, so much the better.

File: src/app/CommandSender.h [hunk 33]
```diff
@@ -36,15 +38,71 @@
...
+         *
+         * @param[in] apCommandSender: The command sender object that initialized the command transaction continue to exist after
+         * this call completes.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.h:64]: OK.  That was not even in the set of possible interpretations of the comment that I considered, so we have clarification work to do here.

That said, we are passing a `const CommandSender *` here.  So it can't be destroyed via the pointer we passed in.  The app _could_ destroy it by matching against other, non-const, state it has, I guess....

If that's what we are trying to avoid, we really need to make this comment clearer, probably with a pointer to the finalization callback as the place where you're allowed to destroy the object.

File: src/app/CommandSender.h [hunk 34]
```diff
@@ -36,15 +38,71 @@
...
+         * @param[in] aProtocolError: The error occurred during the invoke command interaction, will be CHIP_NO_ERROR when the
+         * CommandSender received a valid response from server.
+         */
+        virtual void OnError(const CommandSender * apCommandSender, Protocols::InteractionModel::Status aInteractionModelStatus,
+                             CHIP_ERROR aProtocolError)
+        {}
+
+        /**
+         * OnFinal will be called when CommandSender has finished all the works, and it is safe to release the
+         * allocated CommandSender object if needed.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.h:91]: > This class should not assume how applications are choosing to allocate such objects. 

Just so we're very clear: _allocation_ is not the same thing as _construction and destruction_.  This class could require single-use construction and destruction and implementations could still do whatever they want with the actual _allocation_, via custom `new` operators or placement new or whatnot.

I feel like having re-init semantics complicates both the implementation and consumers, fwiw, with implementation having a more complicated state machine and consumers needing to orchestrate initialization more carefully.  Much simpler if we just stick to "construct the object, use it, destroy it".  Consumers that want to do something interesting with allocation can still do that.

File: src/app/CommandSender.h [hunk 35]
```diff
@@ -36,15 +38,71 @@
...
+         *
+         * This function is marked as must be implemented to make application aware of releasing the object.
+         *
+         * After the command being sent, this function will always be called exactly once for a single command sender.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.h:95]: > I don't see anywhere in here about the caller (which I presume you mean the application here) being able to call Shutdown (which is marked private btw). 

The API documentation for `SendCommandRequest` says, both before and after this PR:

>    // If SendCommandRequest is never called, or the call fails, the API
>    // consumer is responsible for calling Shutdown on the CommandSender.

(And Shutdown is nonexistent, not private, so no, following the API documentation will not compile).

File: src/app/CommandSender.h [hunk 36]
```diff
@@ -66,7 +124,12 @@ class CommandSender : public Command, public Messaging::ExchangeDelegate
                                  System::PacketBufferHandle && aPayload) override;
     void OnResponseTimeout(Messaging::ExchangeContext * apExchangeContext) override;
 
+    // CommandSender does not have "Shutdown" method, the user will destruct it directly.
```
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/CommandSender.h:127]: No, the comment means what it says: there is no `Shutdown` method on this class after this PR.  But there is still API documentation talking about it and requiring consumers to call it.

File: src/app/MessageDef/CommandPath.cpp (hunk 29 above)
[the remark this answers is not in the document]
Comment (bzbarsky-apple) [reply src/app/MessageDef/CommandPath.cpp:51]: You could in fact write code that enforces ordering and handles missing tags without too much trouble.

The real question is this: is this code spec-compliant?  The spec requires aborting interactions on malformed messages.  Is a message with out-or-order tags malformed?  If not, the spec needs to be _really_ clear about that.

File: src/app/CommandSender.h (hunk 34 above)
[the remark this answers is not in the document]
Comment (msandstedt) [reply src/app/CommandSender.h:91]: Agree with @bzbarsky-apple on this one.

**Review by msandstedt — APPROVED** [review]

**Review by bzbarsky-apple — APPROVED** [review]
Approving, with a few more minor nits left.

**Review by yunhanw-google — APPROVED** [review]
